File: Objective-C/admob/RewardedInterstitialExample/RewardedInterstitialExample/ViewController.m
//
//  Copyright 2022 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "ViewController.h"

#import <GoogleMobileAds/GoogleMobileAds.h>

#import "GoogleMobileAdsConsentManager.h"

/// GameState indicates the progress of the game.
typedef NS_ENUM(NSUInteger, GameState) {
  /** Game has not started. */
  GameStateNotStarted = 0,
  /** Game is playing. */
  GameStatePlaying = 1,
  /** Game is paused. */
  GameStatePaused = 2,
  /** Game has ended. */
  GameStateEnded = 3
};

/// The game length.
static const NSInteger kGameLength = 5;

/// The time length before showing ads.
static const NSInteger kAdIntroLength = 3;

@interface ViewController () <GADFullScreenContentDelegate>

/// The privacy settings button.
@property(weak, nonatomic) IBOutlet UIBarButtonItem *privacySettingsButton;

/// The game text.
@property(nonatomic, weak) IBOutlet UILabel *gameText;

/// The play again button.
@property(nonatomic, weak) IBOutlet UIButton *playAgainButton;

/// The text indicating current coin count.
@property(weak, nonatomic) IBOutlet UILabel *coinCountLabel;

/// Starts a new game. Shows a rewarded interstitial if it's ready.
- (IBAction)playAgain:(id)sender;

/// The rewarded interstitial ad.
@property(nonatomic) GADRewardedInterstitialAd *rewardedInterstitialAd;

/// Number of coins the user has earned.
@property(nonatomic) NSInteger coinCount;

/// The countdown timer.
@property(nonatomic) NSTimer *timer;

/// The amount of time left in the game.
@property(nonatomic) NSInteger timeLeft;

/// The state of the game.
@property(nonatomic) GameState gameState;

@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  // Pause game when application is backgrounded.
  [NSNotificationCenter.defaultCenter addObserver:self
                                         selector:@selector(pauseGame)
                                             name:UIApplicationDidEnterBackgroundNotification
                                           object:nil];

  // Resume game when application becomes active.
  [NSNotificationCenter.defaultCenter addObserver:self
                                         selector:@selector(resumeGame)
                                             name:UIApplicationDidBecomeActiveNotification
                                           object:nil];

  __weak __typeof__(self) weakSelf = self;
  [GoogleMobileAdsConsentManager.sharedInstance
      gatherConsentFromConsentPresentationViewController:self
                                consentGatheringComplete:^(NSError *_Nullable consentError) {
                                  __strong __typeof__(self) strongSelf = weakSelf;
                                  if (!strongSelf) {
                                    return;
                                  }

                                  [strongSelf startNewGame];

                                  if (consentError) {
                                    // Consent gathering failed.
                                    NSLog(@"Error: %@", consentError.localizedDescription);
                                  }

                                  if (GoogleMobileAdsConsentManager.sharedInstance.canRequestAds) {
                                    [strongSelf startGoogleMobileAdsSDK];
                                  }

                                  strongSelf.privacySettingsButton.enabled =
                                      GoogleMobileAdsConsentManager.sharedInstance
                                          .isPrivacyOptionsRequired;
                                }];

  // This sample attempts to load ads using consent obtained in the previous session.
  if (GoogleMobileAdsConsentManager.sharedInstance.canRequestAds) {
    [self startGoogleMobileAdsSDK];
  }
}

- (void)startGoogleMobileAdsSDK {
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    // Initialize the Google Mobile Ads SDK.
    [GADMobileAds.sharedInstance startWithCompletionHandler:nil];

    // Request an ad.
    [self loadRewardedInterstitialAd];
  });
}

#pragma mark Game logic

- (void)startNewGame {
  self.gameState = GameStatePlaying;
  self.playAgainButton.hidden = YES;
  self.timeLeft = kGameLength;
  [self updateTimeLeft];
  self.timer = [NSTimer scheduledTimerWithTimeInterval:1.0f
                                                target:self
                                              selector:@selector(decrementTimeLeft:)
                                              userInfo:nil
                                               repeats:YES];
}

- (void)loadRewardedInterstitialAd {
  GADRequest *request = [GADRequest request];
  [GADRewardedInterstitialAd loadWithAdUnitID:@"ca-app-pub-3940256099942544/6978759866"
                                      request:request
                            completionHandler:^(GADRewardedInterstitialAd *ad, NSError *error) {
                              if (error) {
                                NSLog(@"Failed to load rewarded interstitial ad with error: %@",
                                      error.localizedDescription);
                                self.playAgainButton.hidden = NO;
                                return;
                              }
                              self.rewardedInterstitialAd = ad;
                              self.rewardedInterstitialAd.fullScreenContentDelegate = self;
                            }];
}

- (void)updateTimeLeft {
  self.gameText.text = [NSString stringWithFormat:@"%ld seconds left!", (long)self.timeLeft];
}

- (void)decrementTimeLeft:(NSTimer *)timer {
  self.timeLeft--;
  [self updateTimeLeft];
  if (self.timeLeft == 0) {
    [self endGame];
  }
}

- (void)earnCoins:(NSInteger)coins {
  self.coinCount += coins;
  [self.coinCountLabel setText:[NSString stringWithFormat:@"Coins: %ld", (long)self.coinCount]];
}

- (void)pauseGame {
  if (self.gameState != GameStatePlaying) {
    return;
  }
  self.gameState = GameStatePaused;

  // Prevent the timer from firing while app is in background.
  [self.timer invalidate];
  self.timer = nil;
}

- (void)resumeGame {
  if (self.gameState != GameStatePaused) {
    return;
  }
  self.gameState = GameStatePlaying;

  if (!self.timer) {
    self.timer = [NSTimer scheduledTimerWithTimeInterval:1.0f
                                                  target:self
                                                selector:@selector(decrementTimeLeft:)
                                                userInfo:nil
                                                 repeats:YES];
  }
}

- (void)endGame {
  self.gameState = GameStateEnded;
  [self.timer invalidate];
  self.timer = nil;
  [self earnCoins:1];
  __weak ViewController *weakSelf = self;

  __block BOOL adCanceled = NO;

  UIAlertController *alert = [UIAlertController
      alertControllerWithTitle:@"Game Over!"
                       message:[NSString stringWithFormat:@"Watch an ad for 10 more coins. Video "
                                                          @"starting in %ld seconds...",
                                                          (long)kAdIntroLength]
                preferredStyle:UIAlertControllerStyleAlert];

  UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"No, thanks"
                                                         style:UIAlertActionStyleCancel
                                                       handler:^(UIAlertAction *_Nonnull action) {
                                                         adCanceled = YES;
                                                         self.playAgainButton.hidden = NO;
                                                       }];

  [alert addAction:cancelAction];
  [self presentViewController:alert
                     animated:YES
                   completion:^() {
                     ViewController *strongSelf = weakSelf;
                     if (!strongSelf) {
                       return;
                     }

                     dispatch_after(dispatch_time(DISPATCH_TIME_NOW, kAdIntroLength * NSEC_PER_SEC),
                                    dispatch_get_main_queue(), ^{
                                      [self
                                          dismissViewControllerAnimated:YES
                                                             completion:^{
                                                               if (!adCanceled) {
                                                                 [self showRewardedInterstitialAd];
                                                               }
                                                             }];
                                    });
                   }];
}

- (void)showRewardedInterstitialAd {
  if (self.rewardedInterstitialAd) {
    [self.rewardedInterstitialAd
        presentFromRootViewController:self
             userDidEarnRewardHandler:^{
               GADAdReward *reward = self.rewardedInterstitialAd.adReward;

               NSString *rewardMessage =
                   [NSString stringWithFormat:@"Reward received with "
                                              @"currency %@ , amount %ld",
                                              reward.type, [reward.amount longValue]];
               NSLog(@"%@", rewardMessage);
               // Reward the user for watching the
               // video.
               [self earnCoins:reward.amount.longValue];
             }];
  } else {
    NSLog(@"Ad wasn't ready");
  }
}

#pragma mark GADFullScreeContentDelegate implementation

- (void)adWillPresentFullScreenContent:(id<GADFullScreenPresentingAd>)ad {
  NSLog(@"Ad did present full screen content.");
}

- (void)ad:(id)ad didFailToPresentFullScreenContentWithError:(NSError *)error {
  NSLog(@"Ad failed to present full screen content with error %@.", error.localizedDescription);
  self.rewardedInterstitialAd = nil;
  self.playAgainButton.hidden = NO;
}

- (void)adDidDismissFullScreenContent:(id)ad {
  NSLog(@"Ad did dismiss full screen content.");
  self.rewardedInterstitialAd = nil;
  self.playAgainButton.hidden = NO;
}

#pragma mark Rewarded Interstitial button actions

- (IBAction)privacySettingsTapped:(UIBarButtonItem *)sender {
  [self pauseGame];

  [GoogleMobileAdsConsentManager.sharedInstance
      presentPrivacyOptionsFormFromViewController:self
                                completionHandler:^(NSError *_Nullable formError) {
                                  if (formError) {
                                    UIAlertController *alertController = [UIAlertController
                                        alertControllerWithTitle:formError.localizedDescription
                                                         message:@"Please try again later."
                                                  preferredStyle:UIAlertControllerStyleAlert];
                                    UIAlertAction *defaultAction =
                                        [UIAlertAction actionWithTitle:@"OK"
                                                                 style:UIAlertActionStyleCancel
                                                               handler:^(UIAlertAction *action) {
                                                                 [self resumeGame];
                                                               }];

                                    [alertController addAction:defaultAction];
                                    [self presentViewController:alertController
                                                       animated:YES
                                                     completion:nil];
                                  } else {
                                    [self resumeGame];
                                  }
                                }];
}

- (IBAction)playAgain:(id)sender {
  [self startNewGame];
  if (GoogleMobileAdsConsentManager.sharedInstance.canRequestAds) {
    [self loadRewardedInterstitialAd];
  }
}

@end

File: Swift/admanager/AdManagerInterstitialExample/AdManagerInterstitialExample/ViewController.swift
//
//  Copyright (C) 2015 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import GoogleMobileAds
import UIKit

class ViewController: UIViewController, GADFullScreenContentDelegate {

  enum GameState: NSInteger {
    case notStarted
    case playing
    case paused
    case ended
  }

  /// The game length.
  static let gameLength = 5

  /// The interstitial ad.
  var interstitial: GAMInterstitialAd?

  /// The countdown timer.
  var timer: Timer?

  /// The amount of time left in the game.
  var timeLeft = gameLength

  /// The state of the game.
  var gameState = GameState.notStarted

  /// The date that the timer was paused.
  var pauseDate: Date?

  /// The last fire date before a pause.
  var previousFireDate: Date?

  /// Indicates whether the Google Mobile Ads SDK has started.
  private var isMobileAdsStartCalled = false

  /// The container for the game.
  @IBOutlet weak var gameView: UIView!

  /// A spinner indicating whether the app is loading content.
  @IBOutlet weak var loadingSpinner: UIActivityIndicatorView!

  /// The privacy settings button.
  @IBOutlet weak var privacySettingsButton: UIBarButtonItem!

  /// The countdown timer label.
  @IBOutlet weak var gameText: UILabel!

  /// The play again button.
  @IBOutlet weak var playAgainButton: UIButton!

  @IBAction func privacySettingsTapped(_ sender: UIBarButtonItem) {
    pauseGame()

    GoogleMobileAdsConsentManager.shared.presentPrivacyOptionsForm(from: self) {
      [weak self] formError in
      guard let self else { return }
      guard let formError else { return self.resumeGame() }

      let alertController = UIAlertController(
        title: formError.localizedDescription, message: "Please try again later.",
        preferredStyle: .alert)
      alertController.addAction(
        UIAlertAction(
          title: "OK", style: .cancel,
          handler: { _ in
            self.resumeGame()
          }))
      self.present(alertController, animated: true)
    }
  }

  override func viewDidLoad() {
    super.viewDidLoad()

    // Sets up a loading spinner while consent is being gathered.
    let loadingSpinnerWorkItem = DispatchWorkItem {
      self.loadingSpinner.startAnimating()
    }
    // Show spinner if loading takes longer than 1 second.
    DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(1), execute: loadingSpinnerWorkItem)

    // Pause game when application enters background.
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(ViewController.pauseGame),
      name: UIApplication.didEnterBackgroundNotification, object: nil)

    // Resume game when application becomes active.
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(ViewController.resumeGame),
      name: UIApplication.didBecomeActiveNotification, object: nil)

    GoogleMobileAdsConsentManager.shared.gatherConsent(from: self) { [weak self] consentError in
      guard let self else { return }

      loadingSpinnerWorkItem.cancel()
      self.loadingSpinner.stopAnimating()

      // Animate the visibility of the game UI.
      UIView.animate(withDuration: 0.25) {
        self.gameView.alpha = 1
      }

      self.startNewGame()

      if let consentError {
        // Consent gathering failed.
        print("Error: \(consentError.localizedDescription)")
      }

      if GoogleMobileAdsConsentManager.shared.canRequestAds {
        self.startGoogleMobileAdsSDK()
      }

      self.privacySettingsButton.isEnabled =
        GoogleMobileAdsConsentManager.shared.isPrivacyOptionsRequired
    }

    // This sample attempts to load ads using consent obtained in the previous session.
    if GoogleMobileAdsConsentManager.shared.canRequestAds {
      startGoogleMobileAdsSDK()
    }
  }

  private func startGoogleMobileAdsSDK() {
    DispatchQueue.main.async {
      guard !self.isMobileAdsStartCalled else { return }

      self.isMobileAdsStartCalled = true

      // Initialize the Google Mobile Ads SDK.
      GADMobileAds.sharedInstance().start()
      // Request an ad.
      self.loadInterstitial()
    }
  }

  // MARK: - Game Logic

  fileprivate func startNewGame() {
    gameState = .playing
    timeLeft = ViewController.gameLength
    playAgainButton.isHidden = true
    updateTimeLeft()
    timer = Timer.scheduledTimer(
      timeInterval: 1.0,
      target: self,
      selector: #selector(ViewController.decrementTimeLeft(_:)),
      userInfo: nil,
      repeats: true)
  }

  fileprivate func loadInterstitial() {
    GAMInterstitialAd.load(
      withAdManagerAdUnitID: "/6499/example/interstitial",
      request: GAMRequest()
    ) { (ad, error) in
      if let error = error {
        print("Failed to load interstitial ad with error: \(error.localizedDescription)")
        return
      }
      self.interstitial = ad
      self.interstitial?.fullScreenContentDelegate = self
    }
  }

  fileprivate func updateTimeLeft() {
    gameText.text = "\(timeLeft) seconds left!"
  }

  @objc func decrementTimeLeft(_ timer: Timer) {
    timeLeft -= 1
    updateTimeLeft()
    if timeLeft == 0 {
      endGame()
    }
  }

  @objc func pauseGame() {
    if gameState != .playing {
      return
    }
    gameState = .paused

    // Record the relevant pause times.
    pauseDate = Date()
    previousFireDate = timer?.fireDate

    // Prevent the timer from firing while app is in background.
    timer?.fireDate = Date.distantFuture
  }

  @objc func resumeGame() {
    if gameState != .paused {
      return
    }
    gameState = .playing

    // Calculate amount of time the app was paused.
    let pauseTime = (pauseDate?.timeIntervalSinceNow)! * -1

    // Set the timer to start firing again.
    timer?.fireDate = (previousFireDate?.addingTimeInterval(pauseTime))!
  }

  fileprivate func endGame() {
    gameState = .ended
    timer?.invalidate()
    timer = nil

    let alert = UIAlertController(
      title: "Game Over",
      message: "You lasted \(ViewController.gameLength) seconds",
      preferredStyle: .alert)
    let alertAction = UIAlertAction(
      title: "OK",
      style: .cancel,
      handler: { [weak self] action in
        if let ad = self?.interstitial {
          ad.present(fromRootViewController: self!)
        } else {
          print("Ad wasn't ready")
        }
        self?.playAgainButton.isHidden = false
      })
    alert.addAction(alertAction)
    self.present(alert, animated: true, completion: nil)
  }

  // MARK: - Interstitial Button Actions

  @IBAction func playAgain(_ sender: AnyObject) {
    startNewGame()

    if GoogleMobileAdsConsentManager.shared.canRequestAds {
      loadInterstitial()
    }
  }

  // MARK: - GADFullScreenContentDelegate
  func adWillPresentFullScreenContent(_ ad: GADFullScreenPresentingAd) {
    print("Ad will present full screen content.")
  }

  func ad(_ ad: GADFullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error)
  {
    print("Ad failed to present full screen content with error \(error.localizedDescription).")
  }

  func adDidDismissFullScreenContent(_ ad: GADFullScreenPresentingAd) {
    print("Ad did dismiss full screen content.")
  }

  // MARK: - deinit

  deinit {
    NotificationCenter.default.removeObserver(
      self,
      name: UIApplication.didEnterBackgroundNotification, object: nil)
    NotificationCenter.default.removeObserver(
      self,
      name: UIApplication.didBecomeActiveNotification, object: nil)
  }

}
